### PR TITLE
Vulkan: Parallelize GLSL compilation

### PR DESCRIPTION
**Pull request by hrydgard**
Instead of simply directly compiling GLSL to SPIR-V on the main thread using glslang (which, it turns out, is just as expensive as pipeline creation, at least on my machine), we spawn tasks as the needed shaders are discovered, and then the actual pipeline creation blocks on those tasks to finish. This, almost surprisingly, often manages to spread the work out over 10+ threads in heavy scenarios, reducing the length of shader compilation stutters.

There seems to be some minor serialization over memory allocation from within glslang, but still a win. 

Should reduce shader compilation stutters a bit. There's of course lots more to do in this area.

~~Still a draft as there seems to be a leak of shader modules (according to the validation layers) that I can't quite figure out right now.~~  Never mind, fixed, still leaving as draft due to need for more testing.

A next step might be to run multiple CreateGraphicsPipeline in parallel on the pool too instead of having a single "ShaderCompileThread", but there are recommendations to run compiles that use the same shaders on the same thread so some additional logic would be needed (the ShaderCompileThread should probably remain and try to group the pipeline creation requests to the best of its ability as they come in).

The actual win on Android seems pretty small on older devices with 4 cores unfortunately. Probably need to also look into pipeline creation parallelization.

### CODE REVIEW

**Comment by hrydgard** [comment]
Unfortunately, on my Galaxy S8, there seems to be a global lock in pipeline creation, so that creating multiple pipelines in parallel has ~0 benefit :(

Might not be the same on all platforms. But seems like the main way forward for reducing stutters past this will indeed be to try to create less different shaders, branching on uniforms instead.

**Comment by Gamemulatorer** [comment]
So this doesn't help https://github.com/hrydgard/ppsspp/issues/13480?

**Comment by hrydgard** [comment]
It might, a bit yes. I'll get back to this later, have many other things I want to fix...

**Comment by hrydgard** [comment]
Rebased on master. I removed the last commit that parallelizes the actual pipeline creation (will create a separate PR with it later for evaluation), but at least with this, GLSL compilation is done in parallel, which is nice and does reduce first-time shader stutters a bit.

File: GPU/Vulkan/ShaderManagerVulkan.cpp [hunk 1]
```diff
@@ -115,18 +116,20 @@ std::string VulkanFragmentShader::GetShaderString(DebugShaderStringType type) co
...
 	: vulkan_(vulkan), useHWTransform_(useHWTransform), id_(id) {
 	source_ = code;
-	module_ = CompileShaderModule(vulkan, VK_SHADER_STAGE_VERTEX_BIT, source_.c_str());
+	module_ = CompileShaderModuleAsync(vulkan, VK_SHADER_STAGE_VERTEX_BIT, source_.c_str());
 	if (!module_) {
 		failed_ = true;
 		return;
 	} else {
-		VERBOSE_LOG(G3D, "Compiled vertex shader:\n%s\n", (const char *)code);
+		VERBOSE_LOG(G3D, "Compiled fragment shader:\n%s\n", (const char *)code);
```
Comment (unknownbrackets) [inline GPU/Vulkan/ShaderManagerVulkan.cpp:124]: This seems like a lie?

-[Unknown]

Comment (hrydgard) [reply GPU/Vulkan/ShaderManagerVulkan.cpp:124]: Uh, yeah, no idea what happened there.

**Review by unknownbrackets — APPROVED** [review]

**Pull request merged**